fix #15631

n5m opened the pull request:
fix #15631

n5m committed:
trigger valgrind failure on memory leak

Clyybber reviewed (commented):
The tests without `-d:useMalloc` should be removed as they only really test that we *don't* use malloc or that valgrind can only detect leaks via malloc (not something we should test or rely on); we also don't need seperate tests for orc/arc in this case IMO.

n5m commented:
What is the best way to get the Valgrind tests to pass on non-Linux hosts in the pipelines?

One way could be to rewrite the tests to look like
```nim
discard """
   # Testament configs
   # Testament configs
   # Testament configs
"""
when defined(linux) and sizeof(int) == 8:
  # original test code
  discard alloc(1)
else:
  # simulate the expected failed valgrind output
  echo "   definitely lost: 7 bytes in 2 blocks"
  # exit with the same code as --error-exitcode=1
  quit(1)
```
which follows the example set here:
https://github.com/nim-lang/Nim/blob/d6160ef9e4e43a53c54b21d62f9b3115b87cf610/testament/specs.nim#L308
and fakes the output expected in test cases like
https://github.com/nim-lang/Nim/blob/d6160ef9e4e43a53c54b21d62f9b3115b87cf610/tests/valgrind/tleak_arc_useMalloc.nim#L5
but this does not seem like a great solution.

n5m commented:
Oh, this looks like it will be much better:
https://github.com/nim-lang/Nim/blob/d6160ef9e4e43a53c54b21d62f9b3115b87cf610/testament/specs.nim#L317-L325
https://github.com/nim-lang/Nim/blob/d6160ef9e4e43a53c54b21d62f9b3115b87cf610/testament/specs.nim#L339-L341

n5m committed:
remove non-malloc tests

n5m committed:
remove ORC test

is redundant because we already have an ARC test

n5m committed:
only run valgrind tests on 64-bit Linux

n5m committed:
disable freebsd and openbsd

alaviss commented on tests/valgrind/tleak_arc.nim:
```diff
@@ -0,0 +1,14 @@
+cmd: "nim $target --gc:arc -d:useMalloc $options $file"
+exitcode: 1
+outputsub: "   definitely lost: 7 bytes in 2 blocks"
+disabled: "freebsd"
+disabled: "macosx"
+disabled: "openbsd"
+disabled: "windows"
+disabled: "32bit"
```
You don't need that:

https://github.com/nim-lang/Nim/blob/9a110047cbe2826b1d4afe63e3a1f5a08422b73f/testament/specs.nim#L307-L316

Clyybber replied:
We actually do in this case, since otherwise the test will be run without valgrind on these platforms and thus we won't get the error message/exit code.

alaviss replied:
I think we should add a "valgrind: leak" attribute.

Clyybber replied:
Doesn't help here, but had to be done for the failing tests anyways.

Clyybber committed:
Remove tleak_refc

As to not test implementation details (or bug)

Clyybber committed:
Fix test failures by removing redundant test

Since this tests/shoulfail/tvalgrind.nim was specified here to fail this test
itself fails since it will be skipped on non-linux CI

Clyybber committed:
Remove test, reason detailed in the previous commit

Clyybber committed:
Remove redundant disables

Clyybber committed:
Revert removing disables

Clyybber commented:
`tests/arc/tasyncorc.nim` and `tests/arc/thavlak_orc_stress.nim` leak :)

Clyybber committed:
Add and use valgrind: leaks

Clyybber committed:
Fix

Clyybber commented:
Sorry for force-pushing to your branch @n5m , I thought that a rebase would solve the errors.
I now added a `leaks` switch to the valgrind option to accomodate for the two failing tests.
(Test failures unrelated)

n5m commented:
Quite all right, @Clyybber. Happy to have helped with this PR.

I am interested, however, in seeing [tvalgrind.nim](https://github.com/nim-lang/Nim/pull/15632/commits/a2d4245d1494e65707926ea922e6cfd2f1539920#diff-7fbf4dcb13a5825032238f284b2242dbd7ffb2c15ac708683c4ade9403e9c633L1) added back into the codebase. My rationale is that while [tleak_arc.nim](https://github.com/n5m/Nim/blob/13b9699a501f7c94f8b2e9317c9d6d687266196d/tests/valgrind/tleak_arc.nim) tests that memory leaks are found, tvalgrind.nim tests Testament, *itself*. It also nicely parallels the remaining tests in `testament/tests/shouldfail/`, which test Testament's other configuration options.

Would you be opposed to a tvalgrind.nim like this?
```nim
discard """
valgrind: enabled
cmd: "nim $target --gc:arc -d:useMalloc $options $file"
disabled: "freebsd"
disabled: "macosx"
disabled: "openbsd"
disabled: "windows"
disabled: "32bit"
"""

when defined(linux) and sizeof(int) == 8:
  # test that a memory leak is caught by valgrind, but valgrind only
  # runs on 64-bit Linux machines...
  discard alloc(1)
else:
  # ...so on all other OS/architectures, simulate the same exit code
  # that valgrind would have thrown for this test case
  quit(1) # chosen from --error-exit=1
```

Clyybber commented:
@n5m Oh, yeah that looks fine.
